### PR TITLE
#1009 support finding include without extension

### PR DESCRIPTION
**Pull request by jmecosta**
#1009 fix

### CODE REVIEW

**Comment by jmecosta** [comment]
@guwirth this should handle it

File: cxx-squid/src/test/java/org/sonar/cxx/preprocessor/SourceCodeProviderTest.java [hunk 1]
```diff
@@ -118,6 +118,14 @@ public void getting_file_with_filename_and_cwd() {
   }
 
   @Test
+  public void getting_file_with_filename_and_cwd_no_extension() {
+    String cwd = new File("src/test/resources/codeprovider").getAbsolutePath();
+    String path = "source";
```
Comment (jmecosta) [inline cxx-squid/src/test/java/org/sonar/cxx/preprocessor/SourceCodeProviderTest.java]: @guwirth this is the file without extension, correctly lookup up

**Comment by jmecosta** [comment]
yep, i use sonar.cxx.suffixes.headers

**Comment by guwirth** [comment]
Hello @jmecosta,

looking deeper into this my understanding is:
* that we have to add a new test file to https://github.com/SonarOpenCommunity/sonar-cxx/tree/985c0d7849dffe25eff00aa692bcbd4a1d972e63/cxx-squid/src/test/resources/codeprovider ==> source (without extension)
* a new test to https://github.com/SonarOpenCommunity/sonar-cxx/blob/985c0d7849dffe25eff00aa692bcbd4a1d972e63/cxx-squid/src/test/java/org/sonar/cxx/preprocessor/SourceCodeProviderTest.java

```Java
 private final File expected2 = new File(new File("src/test/resources/codeprovider/source").getAbsolutePath());
```

And test like this:

```Java
@Test
  public void getting_file_relpath_without_extension() {
    String baseDir = new File("src/test").getAbsolutePath();
    String dummycwd = "/";
    String path = "source";
    String includeRoot = new File("src/test/resources/codeprovider").getAbsolutePath();

    codeProvider.setIncludeRoots(Arrays.asList(includeRoot), baseDir);
    assertEquals(expected2, codeProvider.getSourceCodeFile(path, dummycwd, true));
    assertEquals(expected2, codeProvider.getSourceCodeFile(path, dummycwd, false));
  }
```
Is my understanding right?

**Comment by jmecosta** [comment]
@guwirth i think the prob is when a file has a extension but not included in the include directive, the current code (without this) handles files without extensions already. the test ive added covers this limitation

File: cxx-squid/src/test/java/org/sonar/cxx/preprocessor/SourceCodeProviderTest.java (hunk 1 above)
Comment (jmecosta) [inline cxx-squid/src/test/java/org/sonar/cxx/preprocessor/SourceCodeProviderTest.java]: @guwirth this is supposed to mimic a include directive, its not the real source file in filesystem

**Comment by guwirth** [comment]
@jmecosta after playing around with VS: the point is that `#include <iostream>` tries to open a file `ìostream` and not `iostream.h`. So your solution must realy try to open a file without extension and not iterate over `sonar.cxx.suffixes.headers`- sorry for my first wrong hint.

**Comment by jmecosta** [comment]
@guwirth this is how its handled now. it first checks if can find filepath (with or witout extension). if it does not find and it does not contain "." then it fallbacks to other extensions

**Comment by guwirth** [comment]
@jmecosta 

>if it does not find and it does not contain "." then it fallbacks to other extensions

But I think this is wrong. Why should it fallback to other extensions?

**Comment by jmecosta** [comment]
its the case of the #include <string> the user reported. where string.h is available however if we dont fallback it will never be lookup

**Comment by guwirth** [comment]
@jmecosta reading #1009 the issue seems to be that we are not able to handle include files without extension. @sreeramk08 is not asking for a fallback? Please @sreeramk08 correct me if I'm wrong.

**Comment by jmecosta** [comment]
@guwirth the current implementation works fine without extension. file.exists() should handle both files with or without extension. or @wichtounet request is not valid for the current implementation. but this implementation is my understanding of the issue

**Comment by wichtounet** [comment]
@jmecosta I dont' think it was me ;P

**Comment by jmecosta** [comment]
@wichtounet  sorry, i meant @sreeramk08

**Comment by jmecosta** [comment]
in fact current implementation is as per standard. visual studio is also unable to find header if extension is not provided. and string header is found as string in include folders not string.h.

**Pull request closed without merging**